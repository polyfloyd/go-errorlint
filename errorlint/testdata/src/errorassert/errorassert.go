package errorassert

import (
	"errors"
	"fmt"
)

type MyError struct{}

func (*MyError) Error() string {
	return "my custom error"
}

type AnotherError struct{}

func (*AnotherError) Error() string {
	return "another error"
}

func doSomething() error {
	return &MyError{}
}

func doSomethingWrapped() error {
	return fmt.Errorf("wrapped: %w", &MyError{})
}

// This should be flagged - direct type assertion
func TypeAssertionDirect() {
	err := doSomething()
	me, ok := err.(*MyError) // want "type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors"
	if ok {
		fmt.Println("got my error:", me)
	}
}

// This should be flagged - type assertion in a simple assignment
func TypeAssertionAssignment() {
	err := doSomething()
	myErr, ok := err.(*MyError) // want "type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors"
	fmt.Println(myErr, ok)
}

// This should be flagged - type assertion in if statement
func TypeAssertionInIf() {
	err := doSomething()
	if me, ok := err.(*MyError); ok { // want "type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors"
		fmt.Println("got my error:", me)
	}
}

// This should be flagged - type switch
func TypeSwitchStatement() {
	err := doSomething()
	switch err.(type) { // want "type switch on error will fail on wrapped errors. Use errors.As to check for specific errors"
	case *MyError:
		fmt.Println("my error")
	case *AnotherError:
		fmt.Println("another error")
	default:
		fmt.Println("unknown error")
	}
}

// This should be flagged - type switch with assignment
func TypeSwitchWithAssignment() {
	err := doSomething()
	switch e := err.(type) { // want "type switch on error will fail on wrapped errors. Use errors.As to check for specific errors"
	case *MyError:
		fmt.Println("my error:", e)
	case *AnotherError:
		fmt.Println("another error:", e)
	default:
		fmt.Println("unknown error:", e)
	}
}

// This should NOT be flagged - using errors.As
func UsingErrorsAs() {
	err := doSomethingWrapped()
	var me *MyError
	if errors.As(err, &me) {
		fmt.Println("got my error:", me)
	}
}

// This should NOT be flagged - non-error type assertion
func NonErrorTypeAssertion() {
	var i interface{} = "hello"
	if s, ok := i.(string); ok {
		fmt.Println(s)
	}
}
