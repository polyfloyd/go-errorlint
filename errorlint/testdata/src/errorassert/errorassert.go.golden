package errorassert

import (
	"errors"
	"fmt"
)

type MyError struct{}

func (*MyError) Error() string {
	return "my custom error"
}

type AnotherError struct{}

func (*AnotherError) Error() string {
	return "another error"
}

func doSomething() error {
	return &MyError{}
}

func doSomethingWrapped() error {
	return fmt.Errorf("wrapped: %w", &MyError{})
}

// This should be flagged - direct type assertion
func TypeAssertionDirect() {
	err := doSomething()
	me := new(MyError)
	ok := errors.As(err, me) // want "type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors"
	if ok {
		fmt.Println("got my error:", me)
	}
}

// This should be flagged - type assertion in a simple assignment
func TypeAssertionAssignment() {
	err := doSomething()
	myErr := new(MyError)
	ok := errors.As(err, myErr) // want "type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors"
	fmt.Println(myErr, ok)
}

// This should be flagged - type assertion in if statement
func TypeAssertionInIf() {
	err := doSomething()
	if me := new(MyError); errors.As(err, me) { // want "type assertion on error will fail on wrapped errors. Use errors.As to check for specific errors"
		fmt.Println("got my error:", me)
	}
}

// This should be flagged - type switch
func TypeSwitchStatement() {
	err := doSomething()
	{
		var err_case_0 *MyError
		var err_case_1 *AnotherError
		switch {
		case errors.As(err, &err_case_0):
			fmt.Println("my error")
		case errors.As(err, &err_case_1):
			fmt.Println("another error")
		default:
			fmt.Println("unknown error")
		}
	}
}

// This should be flagged - type switch with assignment
func TypeSwitchWithAssignment() {
	err := doSomething()
	{
		var e *MyError
		switch {
		case errors.As(err, &e):
			fmt.Println("my error:", e)
		case errors.As(err, &e):
			fmt.Println("another error:", e)
		default:
			fmt.Println("unknown error:", e)
		}
	}
}

// This should NOT be flagged - using errors.As
func UsingErrorsAs() {
	err := doSomethingWrapped()
	var me *MyError
	if errors.As(err, &me) {
		fmt.Println("got my error:", me)
	}
}

// This should NOT be flagged - non-error type assertion
func NonErrorTypeAssertion() {
	var i interface{} = "hello"
	if s, ok := i.(string); ok {
		fmt.Println(s)
	}
}